### PR TITLE
#74/Add local file interceptor and api file decorator

### PR DESCRIPTION
**Pull request by gbudau**
Subtask of #74

### CODE REVIEW

File: transcendence-app/src/user/decorators/api-file.decorator.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+        type: 'object',
+        properties: {
+          [fileName]: {
+            type: 'string',
+            format: 'binary',
+          },
+        },
+      },
+    })(target, propertyKey, descriptor);
+  };
```
Comment (gbudau) [inline transcendence-app/src/user/decorators/api-file.decorator.ts:17]: This decorator is for documenting an endpoint that accepts multipart/form-data using OpenAPI specification
More info here: https://stackoverflow.com/questions/65082784/how-do-i-annotate-an-endpoint-in-nestjs-for-openapi-that-takes-multipart-form-da

File: transcendence-app/src/shared/local-file/local-file.interceptor.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+interface LocalFileInterceptorOptions {
+  fieldName: string;
+  path?: string;
+  fileFilter?: MulterOptions['fileFilter'];
+  limits?: MulterOptions['limits'];
+}
+
+function LocalFileInterceptor(
+  options: LocalFileInterceptorOptions,
+): Type<NestInterceptor> {
```
Comment (gbudau) [inline transcendence-app/src/shared/local-file/local-file.interceptor.ts:18]: TLDR: It's extending the `FileInterceptor` to use a default directory for storing the files and to optionally add a subpath, filter the file by MIME type or limit by file size.

More info about how this works -> https://wanago.io/2021/11/08/api-nestjs-uploading-files-to-server/ under 'Extending the FileInterceptor'

File: transcendence-app/src/user/user.controller.ts [hunk 3]
```diff
@@ -26,6 +27,27 @@ import {
...
+        ),
+        false,
+      );
+    }
+    callback(null, true);
+  },
+  limits: {
+    fileSize: Math.pow(1024, 2), // 1MB
+  },
+});
```
Comment (gbudau) [inline transcendence-app/src/user/user.controller.ts:50]: This is an instance of the `LocalFileInterceptor`, which allows only JPEG files under 1MB.

We save this into a variable to allow importing it from the unit tests

File: transcendence-app/src/user/decorators/api-file.decorator.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import { ApiBody } from '@nestjs/swagger';
+
+export const ApiFile =
+  (fileName = 'file'): MethodDecorator =>
+  (target, propertyKey, descriptor) => {
+    ApiBody({
+      schema: {
+        type: 'object',
+        properties: {
+          [fileName]: {
```
Comment (Damrod) [inline transcendence-app/src/user/decorators/api-file.decorator.ts:10]: I like this way of mapping objects uniquely to data. I wonder if we have more use cases for this.

File: transcendence-app/src/user/user.controller.ts (hunk 3 above)
Comment (Damrod) [reply transcendence-app/src/user/user.controller.ts:50]: Why only jpg? Are we against any other image format? Anyway, we can discuss this later.

**Review by Damrod — APPROVED** [review]
LGTM!
mixin is an interesting concept I had not bumped into...

File: transcendence-app/src/user/user.controller.ts (hunk 3 above)
Comment (gbudau) [reply transcendence-app/src/user/user.controller.ts:50]: We could allow more image formats :+1:

Comment (gbudau) [reply transcendence-app/src/user/user.controller.ts]: Update: I've added a constant with an array of allowed formats. For now it contains `image/jpeg`, `image/png` and `image/webp`. This way, we can easily add/remove other image formats in the future.

**Comment by gbudau** [comment]
Thank you :+1:

**Pull request merged**